### PR TITLE
drivers: apl-ssp: change clock idle state

### PR DESCRIPTION
**Pull request by bkokoszx**
I've change clock idle state to low. Old:
- SCMODE(2) - data driven (rising), data sampled (falling),
  idle state(high)
New:
- SCMODE(1) - data driven (rising), data sampled (falling),
  idle state(low)

Signed-off-by: Bartosz Kokoszko <bartoszx.kokoszko@linux.intel.com>

### CODE REVIEW

**Comment by plbossart** [comment]
Not sure this is correct. This is for the inverted bitclock case, so you drive on rising, sample on falling, not sure why the idle would be low both for normal and inverted cases? If the idle is low in normal, then my initial assumption was that the inversion makes it idle high.

Not sure if it matters either since the bit clock is always ticking...

**Comment by bkokoszx** [comment]
But since we don't set SFCR flag in SSIOC register, the clock isn't always ticking. Am I right?

**Comment by bkokoszx** [comment]
This is logically justified. But still I think, there should be possibility to choose the clock idle state. I can make it using quirk field in sof_ipc_dai_ssp_params struct. What do you think?

**Comment by plbossart** [comment]
Sorry for the delay in answering.
I have no objection to controlling the level, but I'd keep the quirks for register settings. I was thinking we could have a 32-bit clock control field instead of the existing mclk/bclk/fs_keep active and set properties as needed as flags. The idle level would be an additional flag.
Makes sense?

**Comment by bkokoszx** [comment]
@plbossart I've added clks_control field. There are mclk/bclk/fs_keep_active and bclk idle state bits. I've also add handling of idle state. What do you think?

File: src/include/uapi/ipc.h [hunk 1]
```diff
@@ -243,6 +244,20 @@ struct sof_ipc_compound_hdr {
 #define SOF_DAI_INTEL_SSP_FRAME_PULSE_WIDTH_MAX		38
 #define SOF_DAI_INTEL_SSP_SLOT_PADDING_MAX		31
 
+/* SSP clocks control settings
+ *
+ * Macros for clks_control field in sof_ipc_dai_ssp_params struct.
+ */
+
+/* mclk keep active */
```
Comment (plbossart) [inline src/include/uapi/ipc.h]: I would like to add an MCLK_0/1_ENABLE to be able to control the SSP mdivc (or mdivr?) register

File: src/include/uapi/ipc.h [hunk 2]
```diff
@@ -272,9 +287,7 @@ struct sof_ipc_dai_ssp_params {
 
 	/* MCLK */
 	uint32_t mclk_direction;
-	uint32_t mclk_keep_active;
-	uint32_t bclk_keep_active;
-	uint32_t fs_keep_active;
+	uint32_t clks_control;
```
Comment (plbossart) [inline src/include/uapi/ipc.h]: maybe reorder with
frame_pulse_width
tdm_per_slot_padding_flag
clk_control
quirks

File: src/include/uapi/ipc.h [hunk 3]
```diff
@@ -40,6 +40,7 @@
 #define __INCLUDE_UAPI_IPC_H__
 
 #include <uapi/abi.h>
+#include <sof/io.h>
```
Comment (plbossart) [inline src/include/uapi/ipc.h]: why is this needed? To get the BIT() definition?

File: src/drivers/apl-ssp.c [hunk 4]
```diff
@@ -148,6 +148,7 @@ static inline int ssp_set_config(struct dai *dai,
 	uint32_t active_rx_slots = 2;
 	uint32_t sample_width = 2;
 
+	bool inverted_clock = false;
```
Comment (plbossart) [inline src/drivers/apl-ssp.c]: inverted_bclk?

File: src/drivers/apl-ssp.c [hunk 5]
```diff
@@ -242,18 +243,26 @@ static inline int ssp_set_config(struct dai *dai,
...
 		break;
 	default:
 		trace_ssp_error("ec3");
 		ret = -EINVAL;
 		goto out;
 	}
 
+	/* supporting bclk idle state */
+	if (ssp->params.clks_control &
+		SOF_DAI_INTEL_SSP_CLKCTRL_BCLK_IDLE) {
```
Comment (plbossart) [inline src/drivers/apl-ssp.c]: it's really BCLK_IDLE_LOW?

File: src/drivers/apl-ssp.c [hunk 6]
```diff
@@ -242,18 +243,26 @@ static inline int ssp_set_config(struct dai *dai,
...
+		sspsp |= SSPSP_SCMODE((inverted_clock ^ 0x3) & 0x3);
```
Comment (plbossart) [inline src/drivers/apl-ssp.c]: is the ^ 0x3 correct? You are only looking at one bit here. or I am missing the optimization.

File: src/include/uapi/ipc.h (hunk 3 above)
Comment (bkokoszx) [reply src/include/uapi/ipc.h]: Yes, it is for BIT() macro.

File: src/drivers/apl-ssp.c (hunk 6 above)
Comment (bkokoszx) [reply src/drivers/apl-ssp.c]: There is no bit directly responsible for idle state. Clock signal polarity and clock idle state are interrelated. SCMODE in SSPSP can be programmed:
00: Data driven (falling), data sampled (rising), idle state (low)
01: Data driven (rising), data sampled (falling), idle state (low)
10: Data driven (rising), data sampled (falling), idle state (high)
11: Data driven (falling), data sampled (rising), idle state (high)

File: src/drivers/apl-ssp.c (hunk 5 above)
Comment (bkokoszx) [reply src/drivers/apl-ssp.c]: I've added comments.

File: src/include/uapi/ipc.h (hunk 1 above)
Comment (bkokoszx) [reply src/include/uapi/ipc.h]: Do you mean controlling mdivc register insted of:
mdivc |= 0x1;?
I could add this in another patch.

Comment (plbossart) [reply src/include/uapi/ipc.h]: yes that's what I meant. You don't need to enable the outputs on all devices, some don't even have a physical pin.

File: src/drivers/apl-ssp.c (hunk 5 above)
Comment (plbossart) [reply src/drivers/apl-ssp.c]: might as well call the value SOF_DAI_INTEL_SSP_CLKCTRL_BCLK_IDLE_HIGH to make it self-explanatory.

**Comment by plbossart** [comment]
I'd like the addition of the MCLK_0..1_ENABLE bits for a future patch as well as a clarification so that the define BCLK_IDLE_HIGH reflects the functionality.
We can merge on Monday when we have the update.

**Comment by bkokoszx** [comment]
@plbossart I've added MCLK_ENABLE macros and corrected BCLK_IDLE.

**Comment by lgirdwood** [comment]
@bkokoszx ping, is there any update ?

**Comment by bkokoszx** [comment]
I apologize for the delay, I will update this PR on monday.

**Comment by bkokoszx** [comment]
I've changed macros due to @plbossart  suggestions.
